Feature/feature#60 FCM Notification

soochan8 opened the pull request:
### **🔖 요약**
1. FirebaseService 구현 및 FCM 알림 처리 로직 추가

### **🛠 작업 내용**
1. `FirebaseService`
- Notification 생성 및 표시하였습니다.
- 클릭 시, 딥링크 이동 처리하였습니다.
2. `AndroidManifest`
- permission 및 딥링크 인텐트 필터 추가하였습니다.

### **💡 테스트 방법**
- Postman에서 FCM 메시지 전송하여, 알림 수신 및 클릭 이동 확인하였습니다.

### **❓ 궁금 사항**
1. 보통 실무에서 알림 기능을 처리할 때, 이러한 방식으로 처리하는 지 궁금합니다. 다른 방법이나 효율적인 방안이 있다면 여쭤보고 싶습니다.

soochan8 committed:
feat: Manifest permission 설정 및 딥링크 intent-filter 설정

- POST_NOTIFICATIONS 설정

soochan8 committed:
feat: 앱 시작 시, NotificationChannel 추가

soochan8 committed:
feat: MainViewModel 생성 및 딥링크 처리 로직 추가

- 딥링크(coupon) 처리 로직 추가

soochan8 committed:
feat: 딥링크 네비게이션 처리 및 Android 13 알림 권한 요청 추가

- ViewModel의 deepLinkNavigation collect 후, webView 화면 이동

soochan8 committed:
feat: FCM 알림 수신 및 딥링크 이동 기능 추가

- Notification 표시 및 클릭 시 딥링크 화면 이동
- PendingIntent 처리

f-lab-pepe commented on app/src/main/java/com/chan/commerce/MainViewModel.kt:
```diff
@@ -0,0 +1,36 @@
+@HiltViewModel
+class MainViewModel @Inject constructor() : ViewModel() {
+
+    private val _deepLinkNavigation = MutableSharedFlow<String>()
```
MainActivity.onCreate()에서 handleDeepLink(intent)를 먼저 호출하고, 이후 setContent → AppNavHost에서 deepLinkNavigation.collect를 시작하는데 지금 _deepLinkNavigation은 MutableSharedFlow() 기본값(= replay 0)이므로 콜드스타트에서 emit이 수집 시작 전에 발생하면 이벤트가 사라질것 같아요.

soochan8 replied:
넵! SharedFlow는 HotStream이므로 멘토님 말씀처럼 나중에 emit이 된 시점에 collect을 하지 않으므로 이벤트를 받지못해 사라지게됩니다. 해당 부분은 replay = 1을 사용하여 emit 1개를 보장하도록 수정하겠습니다.

f-lab-pepe commented on core/notification/src/main/java/com/chan/notification/data/FirebaseService.kt:
```diff
@@ -0,0 +1,64 @@
+
+
+        val intent = if (!deeplink.isNullOrBlank()) {
+            Intent(Intent.ACTION_VIEW, Uri.parse(deeplink)).apply {
```
스킴검사가 필요할것 같아요. 외부 url을 임의로 실행하면 리스크가 큽니다.

soochan8 replied:
넵 검사 로직 추가하겠습니다! 감사합니다!

soochan8 committed:
feat: deepLink SharedFlow replay 1 설정

- 현재 시점이 emit 후, collect을 하고 있으므로 이벤트를 전달 받을 수 없어 replay 1을 사용하여 최근 이벤트 1개를 전달받을 수 있도록 보장

soochan8 committed:
feat: DeepLinkManager 추가

- 안전한 딥링크 파싱 및 검증
- 허용된 도메인 및 로컬만 접근 허용

soochan8 committed:
feat: mainViewModel 딥링크 파싱 로직 DeepLinkManager 사용

- DeepLinkManager 주입으로 스킴/도메인 검증 책임 분리

soochan8 committed:
feat: FCM 수신 알림 스킴 검증 로직 추가

- 스킴만 빠르게 검증

soochan8 committed:
feat: Hosts 수정

soochan8 committed:
chore: build hilt dependency 추가

soochan8 committed:
chore: intent-filter host 변경

soochan8 commented:
### **🔖 요약**
1. `deepLinkSharedFlow` replay 설정
2. deepLink 스킴 검사 로직

### **🛠 작업 내용**
1. `MainViewModel`
- SharedFlow는 HotStream으로 emit 시점에 collect가 없으면 이벤트가 손실되어 수신받지 못합니다. 이에 따라 `replay = 1`을 설정하여 최근 이벤트 1개를 받도록 보장하였습니다.
2. `DeepLinkManager, DeepLinkDestination`
- URI를 한 곳에 검증하기 위해 DeepLinkManager를 새로 생성하였습니다.
- `DeepLinkManager`의 `parse`를 통해 안전한 스킴 및 도메인 검증을 추가하였습니다.
- `DeepLinkDestination`을 추가하여 이동 경로를 sealed class로 추가하였습니다.

f-lab-pepe commented on app/src/main/java/com/chan/commerce/MainActivity.kt:
```diff
@@ -45,14 +55,42 @@ import javax.inject.Inject
+                    Manifest.permission.POST_NOTIFICATIONS
+                ) != PackageManager.PERMISSION_GRANTED
+            ) {
+                ActivityCompat.requestPermissions(
```
Tiramisu(33) 이상에서 `requestPermissions()` 직접호출은 권장하지 않아요.

```
registerForActivityResult(
    ActivityResultContracts.RequestPermission()
)
```

를 사용해보세요. 그리고 사용자가 허용안했을때 처리가 없어요.
